File: Programs/Lockers.cpp
#include <iostream>
#include <map>

int main() {
	std::map <int, std::string> Lockers ;

	int N_OF_LOCKERS = 100 ; 

	// This creates the dict. of lockers 
	// +1 N_OF_LOCKERS since numbering starts at 1 and would leave 99 lockers.
	int i ;
	for ( i = 1 ; i < N_OF_LOCKERS+1 ; ++i ) {
		Lockers[i] = "closed" ;
	} ;

	// This loop is supposed to continue for 100 times, meanwhile the 
	// 'for' nested in the 'while' loop is to iterate however many times
	// it needs, each time with a step being the number of rounds the 
	// 'while' loop has done, each iteration closing or opening the lockers. 
	int rounds = 0 ;
	while(1) {
		++rounds ;
		
		if (rounds == N_OF_LOCKERS+1) break ;

		for ( i = rounds ; i < N_OF_LOCKERS+1 ; i += rounds ) {
			if (	Lockers[i] == std::string("open") ) {
				Lockers[i] = std::string("closed") ; 
			} else if ( 	Lockers[i] == std::string("closed") ) {
					Lockers[i] = std::string("open") ;
			} ;
		} ;
	} ;

	// This automatically iterate through all keys in dict. using
	// 'locker' to refer to each key.
	// 'first' refers to the key, and 'second' to the value. 
	for ( auto& locker : Lockers ) {
		if ( locker.second == "open" ) {
			std::cout << "Locker #" << locker.first ;
			std::cout << " is " << locker.second << std::endl ;
		} ;
	} ; 

	return 0 ;
} ;

File: Programs/Lockers.py
#!/usr/bin/env python
# Got this problem from TedEd riddles where you get 100 lockers, each with a key.
# The person who gets key 1 will open every locker, person with key 2
# will open or close every other locker and so on. Which lockers will be open
# by the 100th iteration?

num_locks = 100

# First make a dict with 100 (or how many you want) lockers each of them closed.
lockers = {}
for i in range(1, num_locks+1): # since range doesn't count the actual number.
    lockers[i] = "closed"

# Now iterate each one with the rules.
rounds = 0
while(1):
    rounds += 1 

    if rounds == num_locks+1:
        break
    
    for num in range(rounds, num_locks+1, rounds):
        if lockers[num] == "open":
            lockers[num] = "closed"

        elif lockers[num] == "closed":
            lockers[num] = "open"

# Finally iterate through all lockers and check which ones are open.
for key, value in lockers.iteritems():
    if value == "open":
        print("%s is %s" % (key, value) )
